src/laddie/laddie_tracers.f90

```diff
@@ -0,0 +1,296 @@
+  USE mpi_basic                                              , ONLY: par, sync
+  USE control_resources_and_error_messaging                  , ONLY: crash, init_routine, finalise_routine, colour_string
+  USE model_configuration                                    , ONLY: C
+  USE parameters
+  USE mesh_types                                             , ONLY: type_mesh
+  USE ice_model_types                                        , ONLY: type_ice_model
+  USE laddie_model_types                                     , ONLY: type_laddie_model, type_laddie_timestep
+  USE ocean_model_types                                      , ONLY: type_ocean_model
+  USE reallocate_mod                                         , ONLY: reallocate_bounds
+  USE mpi_distributed_memory                                 , ONLY: gather_to_all_dp_1D, gather_to_all_logical_1D
+  USE math_utilities                                         , ONLY: check_for_NaN_dp_1D
+
+  IMPLICIT NONE
+    
+CONTAINS
+    
+! ===== Main routines =====
+! =========================
+
+  SUBROUTINE compute_TS_npx( mesh, ice, laddie, npxref, npx, Hstar, dt, incldiff)
+    ! Integrate T and S by one time step
+
+    ! In- and output variables
+
+    TYPE(type_mesh),                        INTENT(IN)    :: mesh
+    TYPE(type_ice_model),                   INTENT(IN)    :: ice
+    TYPE(type_laddie_model),                INTENT(INOUT) :: laddie
+    TYPE(type_laddie_timestep),             INTENT(IN)    :: npxref
+    TYPE(type_laddie_timestep),             INTENT(INOUT) :: npx
+    REAL(dp), DIMENSION(mesh%vi1:mesh%vi2), INTENT(IN)    :: Hstar
+    REAL(dp),                               INTENT(IN)    :: dt
+    LOGICAL,                                INTENT(IN)    :: incldiff
+
+    ! Local variables:
+    CHARACTER(LEN=256), PARAMETER                         :: routine_name = 'compute_TS_npx'
+    INTEGER                                               :: vi
+    REAL(dp)                                              :: dHTdt
+    REAL(dp)                                              :: dHSdt
+    REAL(dp)                                              :: HT_next
+    REAL(dp)                                              :: HS_next
+ 
+    ! Add routine to path
+    CALL init_routine( routine_name)
+
+    ! Compute divergence of heat and salt
+    CALL compute_divQTS( mesh, laddie, npx, Hstar)
+
+    IF (incldiff) THEN
+      ! Compute RHS including diffusive term
+
+      ! Loop over vertices
+      DO vi = mesh%vi1, mesh%vi2
+        IF (laddie%mask_a( vi)) THEN
+
+          ! == Temperature integration ==
+
+          ! Get dHT_dt
+          dHTdt = -laddie%divQT( vi) &
+                + laddie%melt( vi) * laddie%T_base( vi) &
+                + MAX(0.0_dp,laddie%entr( vi)) * laddie%T_amb( vi) &
+                + laddie%entr_dmin( vi) * laddie%T_amb( vi) &
+                - laddie%detr( vi) * npxref%T( vi) &
+                + laddie%diffT( vi)
```

Comment:
Am I correct that the only difference between the two versions of the calculation is the last term "...+ laddie%diffT( vi)"? If so, maybe cleaner to condense it into a single code version, with a little IF statement in the middle:

dHTdt = ....
if (include_diffusive_terms)
  dHTdt = dHTdt + laddie%diffT( vi)
end if

Reply:
That's how I had it before, but this is the stuff that's called each time step, so I want to minimise all unnecessary statements within do loops.

I'll clean this up